### PR TITLE
gui: Uppercase bech32 addresses in qr codes

### LINKED ISSUE #12191: bech32-optimized QR codes
@sipa said in #11991:

> I haven't looked at the code, but inside QR codes it's advantageous to use uppercase Bech32 (it will result in a smaller QR encoding).

Is that still the case when rendered as URI?

Should amounts be added in a different way?

Would that involve amending BIP-21?

### PR DESCRIPTION
**Pull request by benthecarman**
Closes #12191

### CODE REVIEW

**Comment by sipa** [comment]
Concept ACK

File: src/qt/guiutil.cpp [hunk 1]
```diff
@@ -175,7 +175,8 @@ bool parseBitcoinURI(QString uri, SendCoinsRecipient *out)
 
 QString formatBitcoinURI(const SendCoinsRecipient &info)
 {
-    QString ret = QString("bitcoin:%1").arg(info.address);
+    bool bech_32 = (info.address.startsWith("bc1") || info.address.startsWith("tb1"));
```
Comment (laanwj) [inline src/qt/guiutil.cpp]: I don't like hardcoding this here—don't we have constants to use somewhere?

**Comment by laanwj** [comment]
Concept ACK

File: src/qt/guiutil.cpp (hunk 1 above)
Comment (benthecarman) [reply src/qt/guiutil.cpp]: I looked around and couldn't find any constants, I could add them unless someone else knows if some exist already.

Comment (laanwj) [reply src/qt/guiutil.cpp]: there's nothing on `ChainParams` ?

Comment (sipa) [reply src/qt/guiutil.cpp]: Yeah, you can test whether the address starts with `params.Bech32HRP() + "1"`.

Comment (benthecarman) [reply src/qt/guiutil.cpp]: `params` isn't available in this scope

Comment (sipa) [reply src/qt/guiutil.cpp]: You can get the global params object by calling `Params()`.

**Comment by gmaxwell** [comment]
utACK

File: src/qt/guiutil.cpp [hunk 2]
```diff
@@ -175,7 +175,9 @@ bool parseBitcoinURI(QString uri, SendCoinsRecipient *out)
 
 QString formatBitcoinURI(const SendCoinsRecipient &info)
 {
-    QString ret = QString("bitcoin:%1").arg(info.address);
+    bool bech_32 = info.address.startsWith(QString::fromStdString(Params().Bech32HRP() + "1"));
+
+    QString ret = QString("bitcoin:%1").arg(bech_32? info.address.toUpper() : info.address);
```
Comment (jonasschnelli) [inline src/qt/guiutil.cpp]: nit: space between bech32 and the ? operator.

**Comment by jonasschnelli** [comment]
Concept ACK.
About this implementation: a bit odd that a GUI util function needs to parse addresses in order to differentiate between Bech32 and legacy addresses.

**Comment by benthecarman** [comment]
@jonasschnelli It could be done with parameter but then `ReceiveRequestDialog::update()`, `ReceiveCoinsDialog::copyURI()`, `ReceiveRequestDialog::update(), and `ReceiveRequestDialog::on_btnCopyURI_clicked()` would need a new parameter as well

**Comment by laanwj** [comment]
> About this implementation: a bit odd that a GUI util function needs to parse addresses in order to differentiate between Bech32 and legacy addresses.

Very good point Jonas. It seems something that should happen at another level before passing it to this function. E.g. a flag somewhere to generate the beck32 address as uppercase in the first place.

**Comment by benthecarman** [comment]
@laanwj Are you saying that bech32 addresses should always be presented in uppercase?

**Comment by laanwj** [comment]
No, that's not what I'm saying. But if you need one in uppercase it's probably best to do so closer to the source where it's generated.

**Comment by fanquake** [comment]
Regardless of how it's done, is the end goal here to also have the address inside the QR code uppercase? 
As it's currently implemented (3407b44), we only get an uppercase address in the URI, and not for the other two addresses displayed in the QR code dialog. i.e:

qr code

**Comment by benthecarman** [comment]
I feel like the current implementation makes the most sense.  It would be confusing to have addresses shown primarily in lowercase but when generating a URI it will be shown in uppercase.  Here it just has the address in the URI in uppercase for the optimization and the user never knows that the optimization occurred.

**Comment by hebasto** [comment]
Concept ACK.

**Comment by Sjors** [comment]
Concept ACK. I don't mind too much what the URI looks like, but it doesn't need the same optimization as the QR code, so my light preference is to keep that lower case.

**Comment by gmaxwell** [comment]
reACK

**Comment by meshcollider** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15371/commits/3407b446cc5ec0725c0505e3b933f43d86286ad5

**Comment by hebasto** [comment]
NACK
See: https://github.com/bitcoin/bitcoin/issues/12191#issuecomment-485384845

As [BIP 0021](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki) provides usage of `?`, `=` and `&` characters which do not belong to the alphanumeric set, it is impossible to optimize QR codes by making addresses uppercase, unfortunately.

**Comment by sipa** [comment]
@hebasto As I commented there, I believe yoi are misreading the wikipedia summary.

**Comment by hebasto** [comment]
1. My previous statement regarding impossibility to optimize QR codes by making addresses uppercase is wrong (see [qrencode manual](https://fukuchi.org/works/qrencode/manual/qrencode_8h.html#a4cebc3c670efe1b8866b14c42737fc8f)).

2. It worth to mention [BIP 0021](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki) allows further capitalizing:
> The scheme component ("bitcoin:") is case-insensitive...

3. What are benefits of this PR for users / developers / maintainers?

**Comment by jonasschnelli** [comment]
Re utACK 3407b446cc5ec0725c0505e3b933f43d86286ad5
Though I'm not super happy about the string parsing to detect Bech32,... but probably acceptable. Going to merge this.

**Pull request merged**